common/include/stdlib.h

```diff
@@ -9,6 +15,111 @@ int abs(int i);
+      ├────────────────────────────┤◀┐                                             
+      │               │            │ │  ┌───────────────────────────────────────┐  
+      │     Stack     │            │ │  │Stack base.  As stack usage grows, more│  
+      │               │            │ └──│memory below this address is utilized. │  
+      │               ▼            │    └───────────────────────────────────────┘  
+      │                            │                                               
+      ├ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─│◀─┐  ┌────────────────────────────────────────┐
+      │                            │  │  │Actual lower limit for stack depends on │
+      │                            │  │  │how much stack space your code actually │
+      │                            │  └──│needs.  Generally it can't be determined│
+      │                            │     │             ahead of time.             │
+      │                            │     └────────────────────────────────────────┘
+      ├ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─│◀┐                                             
+      │                            │ │                                             
+      │                            │ │                                             
+      │                            │ │                                             
+      │                            │ │  ┌───────────────────────────────────────┐  
+      │                            │ │  │Heap limit.  You can set this from     │  
+      │            ▲               │ └──│within your program.                   │  
+      │     Heap   │               │    └───────────────────────────────────────┘  
+      │            │               │                                               
+      ├────────────────────────────┤◀┐                                             
+      │                            │ │  ┌───────────────────────────────────────┐  
+      │   Your program is loaded   │ │  │   Heap base pointer.  It's location   │  
+      │           here.            │ │  │depends on how big your program is and │  
+      │                            │ └──│how much memory it allocated in global │  
+      ├────────────────────────────┤    │         and static variables.         │  
+      │                            │    └───────────────────────────────────────┘  
+      │   Reserved lower memory:   │                                               
+      │         zero-page,         │                                               
+      │    hardware stack, etc.    │                                               
+      │                            │                                               
+      │                            │                                               
+0x0000└────────────────────────────┘                                               
+*/
+
+void *malloc(size_t size);
```

Comment:
Please mark these allocator functions as weak, so OS/application developers can substitute their own custom malloc and free. This would in turn allow the standard library and new to automatically pick up the new allocator.

Along these lines, we really should try to move the implementation of the allocator into malloc itself; it'd be very surprising if a user malloc wasn't picked up by operator new() and operator new[].